engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/AtomConstructor.java

```diff
@@ -255,21 +255,34 @@ public String toString() {
    *
    * @return the parent of this atom constructor if it is a variant in a sum type, null otherwise
    */
-  public AtomConstructor getParentType() { return parentType; }
+  public AtomConstructor getParentType() {
+    return parentType;
+  }
 
   public void setParentType(AtomConstructor parent) {
+    CompilerDirectives.transferToInterpreterAndInvalidate();
+    if (parentType != null) throw new IllegalStateException("parent type already initialized");
     parentType = parent;
   }
 
   /**
-   * Gets the list of all variants of this constructor if this is a variant in a sum type.
+   * Gets the array of all variants of this constructor if this is a variant in a sum type.
    *
-   * @return
+   * @return the array of variants
    */
-  public List<AtomConstructor> getVariants() { return variants; }
+  public AtomConstructor[] getVariants() {
+    if (variants == null) throw new IllegalStateException("variants not initialized");
```

Comment:
Removing the `if (variants == null)` check altogether would also work. `variants.clone()` is going to throw `NullPointerException` on its own.

Reply:
I figured it was better to be explicit than have yet another `NullPointerException` source when it inevitably arises. I'd also debated about not cloning the result array and making that user responsibility (which would move that NPE all the way out into user code somewhere) in a quest for micro-optimization, but decided to play it safe for now. This makes that decision a little more easily revertible if we need the speed.